pkg/util/watchdog/watchdog.go

```diff
@@ -7,66 +7,42 @@ import (
 	"github.com/pkg/errors"
 )
 
-type Watchdog interface {
-	Start(stop <-chan struct{})
-}
-
 type SimpleWatchdog struct {
 	NewTicker func() *time.Ticker
 	OnTick    func(context.Context) error
 	OnError   func(error)
 	OnStop    func()
 }
 
-func (w *SimpleWatchdog) Start(stop <-chan struct{}) {
+func (w *SimpleWatchdog) Start(ctx context.Context) {
 	ticker := w.NewTicker()
 	defer ticker.Stop()
 
 	for {
-		ctx, cancel := context.WithCancel(context.Background())
-		// cancel is called at the end of the loop
-		go func() {
-			select {
-			case <-stop:
-				cancel()
-			case <-ctx.Done():
-			}
-		}()
-
-		select {
-		case <-stop:
-		default:
-			if err := w.onTick(ctx); err != nil && !errors.Is(err, context.Canceled) {
+		if err := w.onTick(ctx); err != nil {
+			if !errors.Is(err, context.Canceled) && w.OnError != nil {
 				w.OnError(err)
 			}
 		}
-		cancel()
-
 		select {
 		case <-ticker.C:
-			continue
-		case <-stop:
+		case <-ctx.Done():
 			if w.OnStop != nil {
 				w.OnStop()
 			}
-			// cancel will be called by the above goroutine
 			return
```

Comment:
Just one consideration here:
The golang select runs in random rather than rust marco `select!`, if the `ticker.C` and `ctx.Done` occurred at the same time, it might goes into `<-ticker.C` and runs the for loop again. You can add another `ctx.Done` check inner the `ticker.C` but it looks ugly.

I would prefer to make it in order with priority just like this:
```text
select {
  case <-ctx.Done():
    if w.OnStop != nil {
	    w.OnStop()
    }
    return
  default:
}

select {
  case <-ticker.C:
}
```

If there's something wrong with this, please point it out.

Reply:
Correct. It's a little trickier as you still need to select on the ctx in the second select but I'll improve a bit.

Reply:
PTAL @Icarus9913

Reply:
The new one looks good to me.